# 1. 개요
> * dataset은 370개의 feature로 이루어진 데이터
* 산탄데르 은행에서 캐글에 경연 의뢰한 데이터로 피처 이름은 임의로 처리됨
* TARGET 정보가 1이면 불만족, 0이면 만족
* https://www.kaggle.com/c/santander-customer-satisfaction/data

# 2. 파일 읽기

In [1]:
import pandas as pd

df_train = pd.read_csv('data/santander-customer-satisfaction/train.csv')
df_test = pd.read_csv('data/santander-customer-satisfaction/test.csv')

df_train.head(2)

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
0                      0.0                      0.0  ...   
1                      0.0                      0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3     var38  TARGET  
0                     0.0                     0.0  39205.17       0  
1                     0.0                     0.0  49278.03       0  

[2 rows x 371 columns]

# 3. 데이터 전처리

## 3.1 데이터 확인

In [2]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76020 entries, 0 to 76019
Columns: 371 entries, ID to TARGET
dtypes: float64(111), int64(260)
memory usage: 215.2 MB


In [3]:
df_train.describe()

ID           var3         var15  imp_ent_var16_ult1  \
count   76020.000000   76020.000000  76020.000000        76020.000000   
mean    75964.050723   -1523.199277     33.212865           86.208265   
std     43781.947379   39033.462364     12.956486         1614.757313   
min         1.000000 -999999.000000      5.000000            0.000000   
25%     38104.750000       2.000000     23.000000            0.000000   
50%     76043.000000       2.000000     28.000000            0.000000   
75%    113748.750000       2.000000     40.000000            0.000000   
max    151838.000000     238.000000    105.000000       210000.000000   

       imp_op_var39_comer_ult1  imp_op_var39_comer_ult3  \
count             76020.000000             76020.000000   
mean                 72.363067               119.529632   
std                 339.315831               546.266294   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               12888.030000             21024.810000   

       imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
count             76020.000000             76020.000000   
mean                  3.559130                 6.472698   
std                  93.155749               153.737066   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max                8237.820000             11073.570000   

       imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  ...  \
count             76020.000000             76020.000000  ...   
mean                  0.412946                 0.567352  ...   
std                  30.604864                36.513513  ...   
min                   0.000000                 0.000000  ...   
25%                   0.000000                 0.000000  ...   
50%                   0.000000                 0.000000  ...   
75%                   0.000000                 0.000000  ...   
max                6600.000000              6600.000000  ...   

       saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
count             76020.000000             76020.000000   
mean                  7.935824                 1.365146   
std                 455.887218               113.959637   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max               50003.880000             20385.720000   

       saldo_medio_var33_ult1  saldo_medio_var33_ult3  \
count            76020.000000            76020.000000   
mean                12.215580                8.784074   
std                783.207399              538.439211   
min                  0.000000                0.000000   
25%                  0.000000                0.000000   
50%                  0.000000                0.000000   
75%                  0.000000                0.000000   
max             138831.630000            91778.730000   

       saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
count             76020.000000             76020.000000   
mean                 31.505324                 1.858575   
std                2013.125393               147.786584   
min                   0.000000                 0.000000   
25%                   0.000000                 0.000000   
50%                   0.000000                 0.000000   
75%                   0.000000                 0.000000   
max              438329.220000             24650.010000   

       saldo_medio_var44_ult1  saldo_medio_var44_ult3         var38  \
count            76020.000000            76020.000000  7.602000e+04   

>var3의 최소값이 -999999로 무언가 정상적이지 않다고 생각할 수 있다.

## 3.2 이상치 수정

In [4]:
df_train['var3'].value_counts().sort_index()

var3
-999999      116
 0            75
 1           105
 2         74165
 3           108
           ...  
 228           1
 229           1
 231           1
 235           1
 238           1
Name: count, Length: 208, dtype: int64

In [5]:
df_test['var3'].value_counts().sort_index()

var3
-999999      120
 0            76
 1            99
 2         73962
 3           107
           ...  
 227           2
 229           1
 230           2
 233           1
 238           1
Name: count, Length: 199, dtype: int64

In [6]:
df_train['var3'].value_counts()

var3
 2         74165
 8           138
-999999      116
 9           110
 3           108
           ...  
 63            1
 194           1
 40            1
 57            1
 87            1
Name: count, Length: 208, dtype: int64

> var3를 기준으로 훈련 데이터와 검증데이터에 각각 120개 가량의 데이터가 존재한다. 그리고 바로 다음 수치는 0으로 나타난다. 이는 특정 에러가 발생했을 경우 나타났을 것으로 판단하며(보통 개발 시 절대 나타날 수 없는 값을 에러로 사용하는 경우도 있음으로) 가장 빈도가 많은 2로 대체할 예정이다.(4분위수 대다수에 포진됨)

> -999999가 116개 나타난 것을 볼 수 있다. 여기서는 최빈값으로 값을 대체하자

In [7]:
df_train['var3'] = df_train['var3'].replace(-999999, 2)
df_test['var3'] = df_test['var3'].replace(-999999, 2)

In [8]:
df_train['var3'].value_counts().sort_index()

var3
0         75
1        105
2      74281
3        108
4         86
       ...  
228        1
229        1
231        1
235        1
238        1
Name: count, Length: 207, dtype: int64

In [9]:
df_test['var3'].value_counts().sort_index()

var3
0         76
1         99
2      74082
3        107
4         68
       ...  
227        2
229        1
230        2
233        1
238        1
Name: count, Length: 198, dtype: int64

## 3.3 불필요한 feature 제거

> 첫번째 feature인 ID는 단순 식별자 일 것임으로 예측에 불필요한 데이터가 된다. 

In [10]:
df_train.drop('ID', axis=1, inplace=True)
df_test.drop('ID', axis=1, inplace=True)
df_train.head(2)

var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0     2     23                 0.0                      0.0   
1     2     34                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  ...  \
0                      0.0                      0.0                0.0  ...   
1                      0.0                      0.0                0.0  ...   

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3     var38  TARGET  
0                     0.0                     0.0  39205.17       0  
1                     0.0                     0.0  49278.03       0  

[2 rows x 370 columns]

In [11]:
df_test.head(2)

var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0     2     32                 0.0                      0.0   
1     2     35                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  ...  \
0                      0.0                      0.0                0.0  ...   
1                      0.0                      0.0                0.0  ...   

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   

      var38  
0  40532.10  
1  45486.72  

[2 rows x 369 columns]

In [12]:
df_train.shape

(76020, 370)

In [13]:
df_test.shape

(75818, 369)

# 4. 데이터 분리

In [14]:
y_train = df_train['TARGET']
X_train = df_train.drop('TARGET', axis=1)

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, random_state=10)

# 5. XGBoost

In [16]:
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score, accuracy_score

xgb = XGBClassifier(n_estimators=500, random_state=10)
xgb.set_params(early_stopping_rounds=100, eval_metric='auc')
xgb.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_test, y_test)])

[0]	validation_0-auc:0.83897	validation_1-auc:0.80854
[1]	validation_0-auc:0.84599	validation_1-auc:0.81404
[2]	validation_0-auc:0.85062	validation_1-auc:0.81945
[3]	validation_0-auc:0.85433	validation_1-auc:0.81971
[4]	validation_0-auc:0.86081	validation_1-auc:0.82246
[5]	validation_0-auc:0.86391	validation_1-auc:0.82372
[6]	validation_0-auc:0.86816	validation_1-auc:0.82471
[7]	validation_0-auc:0.87068	validation_1-auc:0.82500
[8]	validation_0-auc:0.87452	validation_1-auc:0.82604
[9]	validation_0-auc:0.87668	validation_1-auc:0.82584
[10]	validation_0-auc:0.87915	validation_1-auc:0.82489
[11]	validation_0-auc:0.88151	validation_1-auc:0.82503
[12]	validation_0-auc:0.88330	validation_1-auc:0.82497
[13]	validation_0-auc:0.88505	validation_1-auc:0.82575
[14]	validation_0-auc:0.88687	validation_1-auc:0.82563
[15]	validation_0-auc:0.88780	validation_1-auc:0.82596
[16]	validation_0-auc:0.88825	validation_1-auc:0.82632
[17]	validation_0-auc:0.89034	validation_1-auc:0.82534
[18]	validation_0-au

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=100,
              enable_categorical=False, eval_metric='auc', feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=500, n_jobs=None,
              num_parallel_tree=None, random_state=10, ...)

In [17]:
xgb.predict_proba(X_test)[:5]

array([[0.998345  , 0.00165499],
       [0.9714408 , 0.02855923],
       [0.82064605, 0.17935397],
       [0.9754388 , 0.02456121],
       [0.98132765, 0.01867233]], dtype=float32)

In [18]:
y_test

9513     0
52084    0
24813    1
12744    0
74796    0
        ..
4895     0
47679    0
52669    0
21953    0
33116    0
Name: TARGET, Length: 19005, dtype: int64

In [19]:
pred = xgb.predict(X_test)

accuracy_score(y_test, pred)

0.9621152328334649

In [20]:
xgb_roc_score = roc_auc_score(y_test, xgb.predict_proba(X_test)[:, 1])
print('ROC AUC 값 : ', xgb_roc_score)

ROC AUC 값 :  0.8263172477579321


In [21]:
xgb_roc_score = roc_auc_score(y_test, xgb.predict_proba(X_test)[:, 0])
print('ROC AUC 값 : ', xgb_roc_score)

ROC AUC 값 :  0.17368267608098847


- https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc?hl=ko
- https://bskyvision.com/1165

> AUC는 이진데이터 구분 시 사용되며 만족 불만족 두 가지에 대한 평가를 하여 사용하였다.

> predict_proba는 확률을 얻게 되며 보통 1이 될 확률로 값을 구하게 된다.

In [22]:
from sklearn.model_selection import GridSearchCV

xgb = XGBClassifier(n_estimators=100)

params = {
    'max_depth':[3, 6],
    'min_child_weight':[1,3],
    'colsample_bytree':[0.5, 0.75]
}

grid_cv = GridSearchCV(xgb, param_grid=params, cv=3)
grid_cv.fit(X_train, y_train, eval_set=[(X_test, y_test)])

print('최적 파라미터 : ', grid_cv.best_params_)
print('최고 AUC :', roc_auc_score(y_test, grid_cv.predict_proba(X_test)[:, 1]))

[0]	validation_0-logloss:0.19343
[1]	validation_0-logloss:0.17472
[2]	validation_0-logloss:0.16388
[3]	validation_0-logloss:0.15676
[4]	validation_0-logloss:0.14926
[5]	validation_0-logloss:0.14582
[6]	validation_0-logloss:0.14192
[7]	validation_0-logloss:0.14052
[8]	validation_0-logloss:0.13835
[9]	validation_0-logloss:0.13682
[10]	validation_0-logloss:0.13625
[11]	validation_0-logloss:0.13530
[12]	validation_0-logloss:0.13518
[13]	validation_0-logloss:0.13464
[14]	validation_0-logloss:0.13421
[15]	validation_0-logloss:0.13383
[16]	validation_0-logloss:0.13363
[17]	validation_0-logloss:0.13347
[18]	validation_0-logloss:0.13341
[19]	validation_0-logloss:0.13325
[20]	validation_0-logloss:0.13323
[21]	validation_0-logloss:0.13329
[22]	validation_0-logloss:0.13324
[23]	validation_0-logloss:0.13318
[24]	validation_0-logloss:0.13321
[25]	validation_0-logloss:0.13319
[26]	validation_0-logloss:0.13336
[27]	validation_0-logloss:0.13333
[28]	validation_0-logloss:0.13344
[29]	validation_0-loglos

> 15분 소요

In [23]:
import numpy as np 

xgb = XGBClassifier(n_estimators=1000, random_state=10, learning_rate=0.02, 
                max_depth=3, min_child_weight=3, colsample_bytree=0.75, reg_alpha=0.03)
xgb.set_params(early_stopping_rounds=200, eval_metric='auc')
xgb.fit(X_train, y_train, eval_set=[(X_train,y_train), (X_test, y_test)])

print('ROC AUC :', roc_auc_score(y_test, xgb.predict_proba(X_test)[:, 1]))

[0]	validation_0-auc:0.78294	validation_1-auc:0.77051
[1]	validation_0-auc:0.78293	validation_1-auc:0.77051
[2]	validation_0-auc:0.80809	validation_1-auc:0.79299
[3]	validation_0-auc:0.81007	validation_1-auc:0.78952
[4]	validation_0-auc:0.81283	validation_1-auc:0.79034
[5]	validation_0-auc:0.81364	validation_1-auc:0.79166
[6]	validation_0-auc:0.81506	validation_1-auc:0.79283
[7]	validation_0-auc:0.81339	validation_1-auc:0.79158
[8]	validation_0-auc:0.81521	validation_1-auc:0.79345
[9]	validation_0-auc:0.81521	validation_1-auc:0.79313
[10]	validation_0-auc:0.81411	validation_1-auc:0.79186
[11]	validation_0-auc:0.81427	validation_1-auc:0.79239
[12]	validation_0-auc:0.81558	validation_1-auc:0.79323
[13]	validation_0-auc:0.81576	validation_1-auc:0.79342
[14]	validation_0-auc:0.81704	validation_1-auc:0.79449
[15]	validation_0-auc:0.81790	validation_1-auc:0.79519
[16]	validation_0-auc:0.81948	validation_1-auc:0.79751
[17]	validation_0-auc:0.81944	validation_1-auc:0.79721
[18]	validation_0-au

> 6분 정도 소요

>LightGBM

In [24]:
from lightgbm import LGBMClassifier

lgbm = LGBMClassifier(n_estimators=500)

lgbm.fit(X_train, y_train, eval_set=[(X_test, y_test)])

lgbm_roc_score = roc_auc_score(y_test, lgbm.predict_proba(X_test)[:,1])
print('ROC AUC 값 : ', lgbm_roc_score)

[LightGBM] [Info] Number of positive: 2290, number of negative: 54725
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.016854 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 14202
[LightGBM] [Info] Number of data points in the train set: 57015, number of used features: 251
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.040165 -> initscore=-3.173769
[LightGBM] [Info] Start training from score -3.173769
ROC AUC 값 :  0.7921709609075842


튜닝

In [25]:
lgbm = LGBMClassifier(n_estimators=500)

params = {
    'num_leaves':[32, 64],
    'max_depth':[128, 160],
    'min_child_samples':[60, 100],
    'subsample': [0.8, 1]
}

grid_cv = GridSearchCV(lgbm, param_grid=params, cv=3)
grid_cv.fit(X_train, y_train, eval_set=[(X_train,y_train), (X_test, y_test)])

print('최적 파라미터 : ', grid_cv.best_params_)
lgbm_score = roc_auc_score(y_test, grid_cv.predict_proba(X_test)[:,1])
print('ROC AUC 값 :', lgbm_score)

[LightGBM] [Info] Number of positive: 1527, number of negative: 36483
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.014460 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 12655
[LightGBM] [Info] Number of data points in the train set: 38010, number of used features: 202
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.040174 -> initscore=-3.173541
[LightGBM] [Info] Start training from score -3.173541
[LightGBM] [Info] Number of positive: 1527, number of negative: 36483
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.009578 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 12906
[LightGBM] [Info] Number of data points in the train set: 38010, number of used features: 205
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.040174 -> initscore=-3.173541


In [52]:
params = {
    'num_leaves':[16, 32],
    'max_depth':[64, 96, 128],
    'min_child_samples':[90, 100, 110],
    'subsample': [0.6, 0.8]
}

grid_cv = GridSearchCV(lgbm, param_grid=params, cv=3)
grid_cv.fit(X_train, y_train, eval_set=[(X_train,y_train), (X_test, y_test)])

print('최적 파라미터 : ', grid_cv.best_params_)
lgbm_score = roc_auc_score(y_test, grid_cv.predict_proba(X_test)[:,1])
print('ROC AUC 값 :', lgbm_score)

최적 파라미터 :  {'max_depth': 64, 'min_child_samples': 90, 'num_leaves': 16, 'subsample': 0.6}
ROC AUC 값 : 0.8100656919774812


> 4분

In [53]:
params = {
    'num_leaves':range(1,11, 3),
    'max_depth':range(1, 60, 10),
    'min_child_samples':range(10, 100, 10),
    'subsample': [0.1, 0.3, 0.5]
}

grid_cv = GridSearchCV(lgbm, param_grid=params, cv=3)
grid_cv.fit(X_train, y_train, eval_set=[(X_train,y_train), (X_test, y_test)])

print('최적 파라미터 : ', grid_cv.best_params_)
lgbm_score = roc_auc_score(y_test, grid_cv.predict_proba(X_test)[:,1])
print('ROC AUC 값 :', lgbm_score)

In [26]:
lgbm = LGBMClassifier(n_estimators=1000, num_leaves=4, max_depth=1, min_child_samples=70, subsample=0.1)
lgbm.fit(X_train, y_train, eval_set=[(X_test, y_test)])
lgbm_roc_score = roc_auc_score(y_test, lgbm.predict_proba(X_test)[:,1])
print('ROC AUC 값 :', lgbm_roc_score)

[LightGBM] [Info] Number of positive: 2290, number of negative: 54725
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.020778 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 13941
[LightGBM] [Info] Number of data points in the train set: 57015, number of used features: 214
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.040165 -> initscore=-3.173769
[LightGBM] [Info] Start training from score -3.173769
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive ga